crates/lang_srv/src/analysis.rs

```diff
@@ -362,38 +461,57 @@ impl AnalyzedDocument {
         Some(GotoDefinitionResponse::Scalar(self.location(range)))
     }
 
-    pub fn format(&self) -> Option<Vec<TextEdit>> {
-        let source = &self.source;
-        let arena = &Bump::new();
-
-        let ast = Ast::parse(arena, source).ok()?;
-        let fmt = ast.fmt();
-
-        if source == fmt.as_str() {
-            None
-        } else {
-            let range = self.whole_document_range();
-            let text_edit = TextEdit::new(range, fmt.to_string().to_string());
-            Some(vec![text_edit])
-        }
+    pub(crate) fn module_url(&self, module_id: ModuleId) -> Option<Url> {
+        self.module()?.module_id_to_url.get(&module_id).cloned()
     }
 
-    pub fn semantic_tokens(&self) -> Option<SemanticTokensResult> {
-        let source = &self.source;
-        let arena = &Bump::new();
-
-        let ast = Ast::parse(arena, source).ok()?;
-        let tokens = ast.semantic_tokens();
+    pub fn completion_items(
+        &self,
+        position: Position,
+        latest_doc: &DocInfo,
+    ) -> Option<Vec<CompletionItem>> {
+        let symbol_prefix = latest_doc.get_prefix_at_position(position);
+        eprintln!(
+            "starting to get completion items for prefix: {:?} docVersion:{:?}",
+            symbol_prefix, latest_doc.version
+        );
+        let len_diff = latest_doc.source.len() as i32 - self.doc_info.source.len() as i32;
 
-        let data = arrange_semantic_tokens(tokens, &self.line_info);
+        //We offset the position because we need the position to be in the correct scope in the most recently parsed version of the source. The quick and dirty method is to just remove the difference in lenght between the source files from the offset. This could cause issues, but is very easy
+        //TODO: this is kind of a hack and should be removed once we can do some minimal parsing without full type checking
+        let mut position = position.to_roc_position(&latest_doc.line_info);
+        position.offset = (position.offset as i32 - len_diff - 1) as u32;
```

Comment:
I cannot remove this without an alternative. 
Firstly, it's extremely rare that you would add to a document in front of the cursor and then try to get completions, you would have to paste some info , then very quickly move you mouse  backwards and try to get a completion.
Secondly, almost all completions are from a document that is different from the LKG, because you just typed a character and triggered completion. Without that it's basically pointless and you may as well just block and wait for the document to be parsed or not have completion. The only scenario I can think of where LKG and the current source actually match is if you partially type something, wait for parsing and an error, then hit the hotkey to trigger completion.

Maybe I don't understand you here?

Reply:
> you may as well just block and wait for the document to be parsed or not have completion

Yes, this is what I'm suggesting.

My understanding of the current implementation is that the LKG can be of any age. If that is true, the LKG can be of any age including a document that is now non-representative of the buffer. For example, the LKG could have been calculated, and then the user did a large delete (that does not parse/check) somewhere, and now is requesting a completion. The LKG is still the document from the first calculation (https://github.com/roc-lang/roc/blob/b31a5c6989ff84c5bc9ebefb54e58927e1fafb00/crates/lang_srv/src/registry.rs#L86-L91), but now the current buffer doesn't line up with it. The offset will be incorrect unless the only edits happened after the cursor's current position.

I am suggesting removing the LKG/current distinction for now and block on analysis for completion. I would rather make the analysis faster, as I'm worried that the effort here will lead to wrong completions without a proper diffing strategy.

Reply:
I understand your concern. However, I would say the PR is pointless without this feature. Analysis of the current document happens much slower than I can type, so a completion never appears fast enough for it to be useful and so you may as well not have it.

Would you be willing to accept a compromise?

Changes to the document across more than one line, and longer than a set length will force a full typecheck.

This will fix your big document change example but still allow you to actually get completions while typing